### ISSUE TITLE
$ref to file not compiled 

### ISSUE THREAD

**Issue by j-d-carmichael** [acrontum/boats-cli#6]
Is it possible to output relative file paths opposed to the #ref that would be seen in the final output?

With a file path reference, I can control + click through the files in my ide, without I cannot.. for example:

`$ref: ../../../components/schema/some/model.yml`

**Comment by p-mcgowan** [comment]
as far as I know, not in the final output - only the intermediate build files - since the final output does not reference files outside of its self and the path depends on where the final output ends up.

you would probably have the best luck if you wrote a small plugin that resolves the root path of the output, but that also depends on where the output is.

so if build/out.yml references `#/components/blah` then you would convert have to resolve it as `../src/`.

**Comment by j-d-carmichael** [comment]
https://github.com/j-d-carmichael/boats/blob/main/srcOA2%2Fpaths%2Fv1%2Fweather%2Fpost.yml#L17

Just to clarify,  so the cli can not output in this format?

**Comment by p-mcgowan** [comment]
wow, I totally spaced and through this was the boats repo.

Now that I understand the question - no, it's not currently possible. If you look into the [generators folder](https://github.com/acrontum/boats-cli/blob/main/src/generators/) it's all hardcoded root refs.

I'd still say fixing your ide is the best bet here, but we can add to the list of nice-to-haves either: 1. post-processing (some way to pass the output files through a script) or 2. figure out if we can resolve back the paths from the filenames that subcommands write to.

However to some degree the comment above still applies - since you typically need to do some post processing by hand on the output anyway (giving better summary / description / model examples), having a script / snippet on your machine to translate root ref wouldn't be the worst.

In summary - will add it as a feature request with a few alternative options in the mean time

**Comment by j-d-carmichael** [comment]
The #ref in any ide will never work as the #ref does not exist until BOATS has compiled it all. The index file is itself empty.

Nice, I will wait for the feature then 😎

**Comment by p-mcgowan** [comment]
> The #ref in any ide will never work as the #ref does not exist until BOATS has compiled it all. The index file is itself empty.

well this is why i suggested the plugin approach. Since the index paths are converted from file paths, it's just a matter of reversing that:

![1725883597](https://github.com/user-attachments/assets/64ee0e3e-f494-415c-b6fc-31251b30b299)

However, not all IDEs are created equal ;)

IDE wars aside, for the less fortunate I took a look already and it will take a bit of tinkering to support this, but it's in the backlog

**Comment by j-d-carmichael** [comment]
I don't want to change how i work there though.. both #ref and file path refs are valid, i prefer the file paths  🤷

**Comment by p-mcgowan** [comment]
will try to get to this this week, but i'd like to make sure there are tests before making too many changes.
I think we will swap the default behaviour here to relative since using root ref reliably would involve reading the indexer trim params, which is a bit more effort than necessary

**Comment by p-mcgowan** [comment]
@j-d-carmichael this should be resolved in 1.0.0

the default now is to use relative paths, since we can guarantee that is correct.

The other behaviour is still possible using `--root-ref=Model`, but it can't be guaranteed to be correct - both because technically it should read off disk for that, and because there is logic around suffix trimming / pluralization I didn't want to duplicate here.

**Comment by j-d-carmichael** [comment]
> [@j-d-carmichael](https://github.com/j-d-carmichael) this should be resolved in 1.0.0
> 
> the default now is to use relative paths, since we can guarantee that is correct.
> 
> The other behaviour is still possible using `--root-ref=Model`, but it can't be guaranteed to be correct - both because technically it should read off disk for that, and because there is logic around suffix trimming / pluralization I didn't want to duplicate here.

noice noice!

I will check it out soon :)
